### DISCUSSION TITLE
Can not insert a widget below MaterialApp and above go_router's Navigator

### DISCUSSION THREAD

**Discussion by whollacsek** [csells/go_router#321] [Q&A]
Hi @csells, according to the docs https://gorouter.dev/navigator-builder

> Sometimes it is necessary to insert a widget above the Navigator, but below MaterialApp/CupertinoApp

> For these purposes, you need to use the navigatorBuilder parameter in the GoRouter constructor. This is similar to the builder parameter in the MaterialApp constructor, but gives access to infrastructure provided by MaterialApp.

However I ran into an issue where I provided a wrapper widget to `navigatorBuilder` while defining my go_router instance, but this wrapper widget does not have access to the MaterialApp context. In this widget I'm calling `showModalBottomSheet` which throws:

```
The following assertion was thrown while handling a gesture:
Navigator operation requested with a context that does not include a Navigator.
The context used to push or pop routes from the Navigator must be that of a widget that is a
descendant of a Navigator widget.
When the exception was thrown, this was the stack:
#0      Navigator.of.<anonymous closure> (package:flutter/src/widgets/navigator.dart:2553:9)
#1      Navigator.of (package:flutter/src/widgets/navigator.dart:2560:6)
#2      showModalBottomSheet (package:flutter/src/material/bottom_sheet.dart:694:46)
...
```

My go_router instance is defined as:
```dart
final _router = GoRouter(
  routes: [...],
  navigatorBuilder: (context, state, child) {
    return MyWrapperWidget(child: child);
  },
);
```
And the `MyWrapperWidget` displays on all routes a button that when clicked should show a bottom sheet modal.

How can I have a wrapper around all pages that can access the Navigator from MaterialApp?

**Comment by csells** [comment]
This is the expected behavior, since you've explicitly chosen to add content about the go_router-provided `Navigator` instance. If you'd like to provide you own `Navigator` for this kind of thing, you can do so as shown in [the shared scaffold sample](https://github.com/csells/go_router/blob/main/go_router/example/lib/shared_scaffold.dart#L38).

**Reply by whollacsek** [reply]
But is there a way to use the Navigator from MaterialApp? I don't think having a separate Navigator is a good thing

**Reply by csells** [reply]
The `Navigator` provided by the `MaterialApp` is overridden when you use the `MaterialApp.router` API. go_router provides the specific `Navigator` instance as an implementation of that API. You can add your own `Navigator` at the top of such things if you like, as the sample shows.

**Reply by whollacsek** [reply]
Thanks for the explanation @csells, however now with a `Navigator` added in the `navigatorBuilder`, my app is throwing a framework error:

> 'package:flutter/src/widgets/navigator.dart': Failed assertion: line 3465 pos 14:
'observer.navigator == null': is not true.

Would you have any idea how I could debug/solve this error?

Error stack contains only reference to Flutter framework, `UncontrolledProviderScope` is from [Riverpod source file](https://github.com/rrousselGit/river_pod/blob/43208b8c6a022afad0801ce1375eeb328c7765d3/packages/flutter_riverpod/lib/src/framework.dart#L205) :
```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown building MyWidget(dependencies: [UncontrolledProviderScope],
state: _ConsumerState#14a86):
'package:flutter/src/widgets/navigator.dart': Failed assertion: line 3465 pos 14:
'observer.navigator == null': is not true.
Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md
The relevant error-causing widget was:
  MyWidget
  MyWidget:file:///REDACTED/main.dart:177:35
When the exception was thrown, this was the stack:
#2      NavigatorState.activate (package:flutter/src/widgets/navigator.dart:3465:14)
#3      StatefulElement.activate (package:flutter/src/widgets/framework.dart:4801:11)
#4      Element._activateRecursively (package:flutter/src/widgets/framework.dart:3726:13)
#5      Element._activateWithParent (package:flutter/src/widgets/framework.dart:3719:5)
...     Normal element mounting (9 frames)
#14     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3631:14)
#15     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6261:36)
#16     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6272:32)
...     Normal element mounting (185 frames)
#201    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3631:14)
#202    MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6261:36)
#203    MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6272:32)
...     Normal element mounting (52 frames)

... stack continues...
```

**Reply by csells** [reply]
I've seen this before but I be darned if I remember what caused it or how I fixed it.

@lulupointu @esdotdev @rydmike Does this sound familiar to you?

**Reply by esDotDev** [reply]
Pretty sure this should work fine, can you post a basic reproduction snippet?

**Reply by whollacsek** [reply]
Thanks @esDotDev I'll prepare a repo today and hopefully you'll be able to guide me through it

**Reply by whollacsek** [reply]
@esDotDev @csells I've made the repo https://github.com/whollacsek/go_router_navigator_crash the crash happens when the Navigator is injected dynamically in `navigatorBuilder`. The important things are in `main.dart` and `wrapper.dart`. Let me know if you need me to clarify things

**Reply by esDotDev** [reply]
Looks like you're trying to wrap GoRouters' navigator (child), in a second Navigator, that sometimes disappears from the tree. Not sure why it is blowing up, but that seems like maybe some bug in Navigator which is probably nothing GoRouter can control.

Instead of conditionally adding the navigator, keep it in the tree, and just modify the child instead? 
```dart
bool loggedIn = container.read(loginInfoProvider).loggedIn;
if(loggedIn) child = Authenticated(child: child);
return Navigator(
  onPopPage: (route, dynamic result) {
    route.didPop(result);
    return false; // don't pop the single page on the root navigator
  },
  pages: [MaterialPage<void>(child: child)],
)
```

**Reply by whollacsek** [reply]
I see, I'll go with your suggestion. I wanted to have a semantically correct tree, but seems like for now I'll have to go with this, thanks for taking the time @esDotDev @csells
